tikv-client/src/main/java/com/pingcap/tikv/types/JsonType.java

```diff
@@ -252,69 +211,45 @@ private long parseDataLength(DataInput di) {
   }
 
-  private byte readByte(DataInput cdi) {
-    byte type;
-    try {
-      type = cdi.readByte();
-    } catch (IOException e) {
-      throw new UncheckedIOException(e);
-    }
-    return type;
-  }
-
-  private JsonArray parseArray(DataInput di) {
-    long elementCount = parseUint32(di);
-    long size = parseUint32(di);
-    byte[] buffer = new byte[Math.toIntExact(size - 8)];
-    readFully(di, buffer);
+  private JsonArray parseArray(CodecDataInput cdi, int elementCount, int length) {
+    byte[] buffer = new byte[length];
+    cdi.readFully(buffer);
     JsonArray jsonArray = new JsonArray();
     for (int i = 0; i < elementCount; i++) {
-      JsonElement value = parseValueEntry(buffer, VALUE_ENTRY_SIZE * i);
+      JsonElement value = parseValueEntry(buffer, i * VALUE_ENTRY_SIZE);
       jsonArray.add(value);
     }
     return jsonArray;
   }
 
-  private JsonObject parseObject(DataInput di) {
-    long elementCount = parseUint32(di);
-    long size = parseUint32(di);
-
-    byte[] buffer = new byte[Math.toIntExact(size - 8)];
-    readFully(di, buffer);
+  private JsonObject parseObject(CodecDataInput cdi, int elementCount, int length) {
+    byte[] buffer = new byte[length];
+    cdi.readFully(buffer);
     JsonObject jsonObject = new JsonObject();
+    long valueEntryOffset = elementCount * KEY_ENTRY_LENGTH;
     for (int i = 0; i < elementCount; i++) {
-      KeyEntry keyEntry = parseKeyEntry(ByteStreams.newDataInput(buffer, i * KEY_ENTRY_LENGTH));
-      String key =
-          parseString(
-              ByteStreams.newDataInput(buffer, Math.toIntExact(keyEntry.keyOffset - 8)),
-              keyEntry.keyLength);
-      long valueEntryOffset = elementCount * KEY_ENTRY_LENGTH + i * VALUE_ENTRY_SIZE;
-      JsonElement value = parseValueEntry(buffer, valueEntryOffset);
+      KeyEntry keyEntry = parseKeyEntry(new CodecDataInput(buffer, i * KEY_ENTRY_LENGTH));
+      String key = parseString(new CodecDataInput(buffer, keyEntry.keyOffset), keyEntry.keyLength);
+      JsonElement value = parseValueEntry(buffer, valueEntryOffset + i * VALUE_ENTRY_SIZE);
       jsonObject.add(key, value);
     }
     return jsonObject;
   }
 
   private JsonElement parseValueEntry(byte[] buffer, long valueEntryOffset) {
     byte valueType = buffer[Math.toIntExact(valueEntryOffset)];
-    JsonElement value;
-    ByteArrayDataInput bs = ByteStreams.newDataInput(buffer, Math.toIntExact(valueEntryOffset + 1));
-    switch (valueType) {
-      case TYPE_CODE_LITERAL:
-        value = parseLiteralJson(bs);
-        break;
-      default:
-        long valueOffset = parseUint32(bs);
-        value =
-            parseValue(
-                valueType, ByteStreams.newDataInput(buffer, Math.toIntExact(valueOffset - 8)));
+    CodecDataInput bs = new CodecDataInput(buffer, Math.toIntExact(valueEntryOffset + 1));
+    if (valueType == TYPE_CODE_LITERAL) {
+      return parseLiteralJson(bs);
+    } else {
+      int valueOffset = parseUint32(bs) - 8;
```

Comment:
maybe use a final static variable to describe the magic number `8`

Reply:
agree.